hnn_core/parallel.py

```diff
@@ -7,20 +7,76 @@
 
 from neuron import h
 
-rank = 0
-nhosts = 1
-pc = h.ParallelContext(nhosts)
-pc.done()
-rank = int(pc.id())
-cvode = h.CVode()
+pc = None
+last_network = None
 
 
-def create_parallel_context(n_jobs=1):
-    """Create parallel context."""
-    rank = int(pc.id())     # rank or node number (0 will be the master)
+def shutdown():
+    """
+    Tell nrniv processes to terminate. Calling script will also terminate
+    """
+    pc.done()
+    h.quit()
 
-    if rank == 0:
-        pc.gid_clear()
+
+def get_nhosts():
+    """Return the number of processors used by ParallelContext
+    Returns
+    -------
+    nhosts: int
+        Value from pc.nhost()
+    """
+    return nhosts
+
+
+def create_parallel_context(n_cores=None):
+    """Create parallel context.
+    Parameters
+    ----------
+    n_cores: int | None
+        Number of processors to use for a simulation. A value of None will
+        allow NEURON to use all available processors.
+    """
+
+    global rank, nhosts, cvode, pc, last_network
```

Comment:
same comment as elsewhere, if we can avoid `global` it's better

Reply:
I don't think this can be avoided. It's important for this scheme (which allows multiple Network instatiations) to keep this state. Open to alternative methods, but pc gets called all over, and last_network is internal state that probably shouldn't be exposed to the user if possible.

Reply:
@jasmainak I don't have a better option for this.

Reply:
global variables have a more limited usage now, mainly in `neuron.py`